### ISSUE TITLE
Plans for `Sock.Router`?

### ISSUE THREAD

**Issue by moogle19** [mtrudel/sock#1]
Do you have any plans for adding a router to sock yet?
This is probably the last big obstacle to have Phoenix run on top of bandit.

**Comment by mtrudel** [comment]
As of this week Bandit runs Phoenix, full stop. Check out https://github.com/phoenixframework/phoenix/pull/4973 for details.

I hadn't actually thought about the notion of a `Sock.Router`! I don't know how necessary it is at the moment given that Sock support in Phoenix is currently sort of 'inside-out' from how it ought to be implemented as a necessity of keeping the changes to Phoenix contained in scope (basically, we need to rely on the `Phoenix.Endpoint` socket DSL to do the equivalent of routing for us, since Phoenix needs to support legacy Cowboy & long polling endpoints as well as Sock). I'm still putting together a longer term proposal for all this, but roughly I'm thinking it looks like this:

1. Add `Sock` support to Phoenix, allowing it to be run within a Sock server (this is done in the above PR)
2. Update Plug.Cowboy to support Sock & add a PhoenixAdapter to Plug.Cowboy that uses it
3. Update Phoenix.Endpoint to use this new Plug.Cowboy.PhoenixAdapter in place of Phoenix's internal Cowboy2Adapter
4. Phoenix is now entirely served via Plug & Sock. We can now safely refactor Phoenix.Endpoint's internal `__handler__` mechanism to more explicitly consider Plug & Sock handlers (extra wrinkle, `socket` declarations may be backed by either / both WebSocket or HTTP long polling implementations).

Step 4 is where the idea of a `Sock.Router` comes into play. I'm loosely thinking that Phoenix.Endpoint would build routes for both Plug and Sock at compile time, basically in a more general way than it does today at / around [here](https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/endpoint.ex#L627-L635). Or something like that, anyway.

Long story short, this is very much a thing that makes sense (and thank you for bringing it up!), but it's not coming anytime soon I don't think, and isn't necessary for Phoenix support today.

**Comment by mtrudel** [comment]
Plans have changed for this after consulting with Jose. See https://github.com/phoenixframework/phoenix/pull/4973 for details, but Sock won't have any need for a router because it's being narrowed in scope to only cover 'post-upgrade' concerns. All routing within e.g. Phoenix will be done by Plug.Router while the connection is still HTTP
